### PR TITLE
Swarm particle optimization with  continue_as_new

### PR DESCRIPTION
**Pull request by marcofiocco**

### CODE REVIEW

File: cmd/samples/pso/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+2) Run the following command multiple times on different console window. This is to simulate running workers on multiple different machines.
+```
+./bin/fileprocessing -m worker
+```
+3) Run the following command to submit a start request for this fileprocessing workflow.
+```
+./bin/fileprocessing -m trigger
+```
+
+You should see that all activities for one particular workflow execution are scheduled to run on one console window.
```
Comment (mfateev) [inline cmd/samples/pso/README.md]: Rewrite README to match the sample.

File: cmd/samples/pso/main.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/uber/cadence-samples/cmd/samples/common"
+	"go.uber.org/cadence/client"
+	"go.uber.org/cadence/encoded"
+	"go.uber.org/cadence/worker"
+)
+
+// myDataConverter implements encoded.DataConverter using gob for Swarm and Particle
+// WARGNING: Make sure all struct members are public (Capital letter) otherwise serialization does not work!
+// TODO: consider storing blobs in external DB or Forge
+type myDataConverter struct {
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: May be rename it to something like gobDataConverter?

File: cmd/samples/pso/swarm.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		// Update particles in parallel
+		for i := 0; i < swarm.Settings.Size; i++ {
+			particleIdx := i
+			workflow.Go(ctx, func(ctx workflow.Context) { // Use an activity for this whole block
+				var particle Particle
+				err := workflow.ExecuteActivity(ctx, updateParticleActivityName, *swarm, particleIdx).Get(ctx, &particle)
+				if err == nil {
+					swarm.Particles[particleIdx] = &particle
+				} else {
+					//FATAL ERROR
```
Comment (mfateev) [inline cmd/samples/pso/swarm.go]: Return the error from this function to fail the workflow.

File: cmd/samples/pso/workflow.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+// This is registration process where you register all your workflow handlers.
+func init() {
+	//workflow.Register(FitnessEvaluationWorkflow)
+	workflow.Register(PSOWorkflow)
+	workflow.Register(PSOChildWorkflow)
+}
+
+//PSOWorkflow workflow decider
+func PSOWorkflow(ctx workflow.Context, functionName string) (err error) {
+	logger := workflow.GetLogger(ctx)
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: consider adding queries to report current state of the simulation

File: cmd/samples/pso/README.md (hunk 1 above)
Comment (marcofiocco) [reply cmd/samples/pso/README.md]: Yup

File: cmd/samples/pso/main.go (hunk 2 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: OK

File: cmd/samples/pso/swarm.go (hunk 3 above)
Comment (marcofiocco) [reply cmd/samples/pso/swarm.go]: right

File: cmd/samples/pso/workflow.go (hunk 4 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: Great idea

File: cmd/samples/pso/activities.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	updateParticleActivityName = "updateParticleActivityName"
+)
+
+var rng *rand.Rand
+
+// This is registration process where you register all your activity handlers.
+func init() {
+	// initialize the RNG
+	// WARNING: the randomness of activity scheduling with multiple workers makes random number generation truly random and not repeatable in debugging
+	rng = rand.New(rand.NewSource(time.Now().UnixNano()))
```
Comment (mfateev) [inline cmd/samples/pso/activities.go]: Add comment that worker.ReplayWorkflowHistoryFromJSONFile should be used to troubleshoot a specific workflow failure.

File: cmd/samples/pso/main.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+// This needs to be done as part of a bootstrap step when the process starts.
+// The workers are supposed to be long running.
+func startWorkers(h *common.SampleHelper) {
+	// Configure worker options.
+	workerOptions := worker.Options{
+		MetricsScope:          h.Scope,
+		Logger:                h.Logger,
+		EnableLoggingInReplay: true,
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: I would leave it as a default "false".

File: cmd/samples/pso/main.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		DataConverter:         h.DataConverter,
+	}
+	h.StartWorkers(h.Config.DomainName, ApplicationName, workerOptions)
+
+	// Host Specific activities processing case
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: Why is the host specific task list meeded if there is no caching the result of the previous step?

File: cmd/samples/pso/main.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	workerOptions.DisableWorkflowWorker = true
+	h.StartWorkers(h.Config.DomainName, HostID, workerOptions)
+}
+
+func startWorkflow(h *common.SampleHelper, functionName string) {
+	workflowOptions := client.StartWorkflowOptions{
+		ID:                              "PSO_" + uuid.New(),
+		TaskList:                        ApplicationName,
+		ExecutionStartToCloseTimeout:    time.Minute * 60,
+		DecisionTaskStartToCloseTimeout: time.Minute, // measure of responsiveness of the worker to various server signals apart from start workflow
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: I would change it to 10 seconds

File: cmd/samples/pso/workflow.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		InitialInterval:          time.Second,
+		BackoffCoefficient:       2.0,
+		MaximumInterval:          time.Minute,
+		ExpirationInterval:       time.Minute * 10,
+		MaximumAttempts:          5,
+		NonRetriableErrorReasons: []string{"bad-error"},
+	},
+}
+
+var QueryResult string
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: Cannot use a mutable global variable inside the workflow code as multiple workflow instances might be executed on the same host. Either pass it through the appropriate structures or inside the workflow.Context named variable.

File: cmd/samples/pso/workflow.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	childID := fmt.Sprintf("PSO_child_workflow:%v", execution.RunID)
+	cwo := workflow.ChildWorkflowOptions{
+		// Do not specify WorkflowID if you want cadence to generate a unique ID for child execution
+		WorkflowID:                   childID,
+		ExecutionStartToCloseTimeout: time.Minute,
+	}
+	ctx = workflow.WithChildOptions(ctx, cwo)
+
+	// setup query handler for query type "state"
+	QueryResult = "started"
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: Cannot use global variables to share data across multipel workflow instances. The solution is to have two queries. One in the parent workflow to return the id of the currently running child the other one is in the child to return its current status.
It currently works only because a single worker process is used. As soon as this workflow is deployed to a cluster the query is going to break.

File: cmd/samples/pso/workflow.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+			return false, workflow.NewContinueAsNewError(ctx, PSOChildWorkflow, swarm, result.Step+1)
+		}
+
+		logger.Error("Error in swarm loop: ", zap.Error(err))
+		return false, errors.New("Error in swarm loop")
+	}
+	if result.Position.Fitness < swarm.Settings.Function.Goal {
+		msg := fmt.Sprintf("Yay! Goal was reached @ step %d (fitness=%.2e) :-)", result.Step, result.Position.Fitness)
+		logger.Info(msg)
+		return true, nil
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: I"m confused about the result of this workflow. Does it save the result of the computation anywhere?

File: cmd/samples/pso/swarm.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+		if step == swarm.Settings.Steps {
+			break
+		}
+
+		// Not finished yet, just continue as new to reduce history size
+		if step%swarm.Settings.ContinueAsNewEvery == 0 {
+			return Result{
+				Position: *swarm.Gbest,
+				Step:     step,
+			}, errors.New("CONTINUEASNEW")
```
Comment (mfateev) [inline cmd/samples/pso/swarm.go]: Probably worth defining this error as a constant.

File: cmd/samples/pso/main.go [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/uber/cadence-samples/cmd/samples/common"
+	"go.uber.org/cadence/client"
+	"go.uber.org/cadence/encoded"
+	"go.uber.org/cadence/worker"
+)
+
+// gobDataConverter implements encoded.DataConverter using gob for Swarm and Particle
+// WARGNING: Make sure all struct members are public (Capital letter) otherwise serialization does not work!
+// TODO: consider storing blobs in external DB or S3
+type gobDataConverter struct {
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: I think it is worth moving into its own file

File: cmd/samples/pso/main.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+// This needs to be done as part of a bootstrap step when the process starts.
+// The workers are supposed to be long running.
+func startWorkers(h *common.SampleHelper) {
+	// Configure worker options.
+	workerOptions := worker.Options{
+		MetricsScope:          h.Scope,
+		Logger:                h.Logger,
+		EnableLoggingInReplay: true,
+		DataConverter:         h.DataConverter,
+	}
```
Comment (mfateev) [inline cmd/samples/pso/main.go]: I think you want to set MaxConcurrentActivityExecutionSize option to 1 as the activities are CPU intensive.
I think putting a comment on this value would also be usefult.

File: cmd/samples/pso/activities.go (hunk 5 above)
Comment (marcofiocco) [reply cmd/samples/pso/activities.go]: Cool, I didn't know about that feature

File: cmd/samples/pso/main.go (hunk 6 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Agreed

File: cmd/samples/pso/main.go (hunk 7 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Oh, that was a leftover from the initial sample I was taking inspiration from

File: cmd/samples/pso/main.go (hunk 8 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Ok, but why? What's the benefit?

File: cmd/samples/pso/workflow.go (hunk 9 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: I've tried to define 
```
err = workflow.SetQueryHandler(ctx, "state", func(input []byte) (string, error) {
	return ctx.Value("QueryResult").(string), nil
})
```
and then set the values like
```
ctx = workflow.WithValue(ctx, "QueryResult", "initialized")
```
but I have problems reading the latest result, it seems is not updated. Do I have to set the BackgroundActivityContext like in the recovery sample?

File: cmd/samples/pso/workflow.go (hunk 10 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: Ok, maybe I got it, see my comment above too

File: cmd/samples/pso/workflow.go (hunk 11 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: Not at the moment, it's just saying that the optimization has converged (true)

File: cmd/samples/pso/swarm.go (hunk 12 above)
Comment (marcofiocco) [reply cmd/samples/pso/swarm.go]: Agreed

File: cmd/samples/pso/main.go (hunk 14 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Good point

File: cmd/samples/pso/main.go (hunk 13 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Good idea

File: cmd/samples/pso/main.go (hunk 8 above)
Comment (mfateev) [reply cmd/samples/pso/main.go]: Faster recovery in the case of worker failure.

File: cmd/samples/pso/workflow.go (hunk 9 above)
Comment (mfateev) [reply cmd/samples/pso/workflow.go]: You cannot share variables across workflows. As you set handler in a parent and update the value in the child it is not going to work.

File: cmd/samples/pso/workflow.go (hunk 11 above)
Comment (mfateev) [reply cmd/samples/pso/workflow.go]: May be add comment to make it clear.

File: cmd/samples/pso/main.go (hunk 8 above)
Comment (marcofiocco) [reply cmd/samples/pso/main.go]: Right

File: cmd/samples/pso/workflow.go (hunk 11 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: ok

File: cmd/samples/pso/workflow.go (hunk 9 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: I see. I've tried but I have not figure out how to get the RunId of the child workflow from the parent workflow, I think it's necessary with the ChildID to create a query.

Also I cannot figure out how to make a query to the child workflow in the test environment.

File: cmd/samples/pso/workflow.go [hunk 15]
```diff
@@ -68,33 +66,34 @@ func PSOWorkflow(ctx workflow.Context, functionName string) (err error) {
...
 			return err
 		}
 
-		QueryResult = "initialized"
-
 		var goalReached bool
-		err = workflow.ExecuteChildWorkflow(ctx, PSOChildWorkflow, *swarm, 1).Get(ctx, &goalReached)
+		childFuture := workflow.ExecuteChildWorkflow(ctx, PSOChildWorkflow, *swarm, 1)
+		//var childWE WorkflowExecution
+		//childFuture.GetChildWorkflowExecution().Get(&childWE)
```
Comment (marcofiocco) [inline cmd/samples/pso/workflow.go]: Cannot get the child WorkflowExecution and RunID here.

File: cmd/samples/pso/workflow_test.go [hunk 16]
```diff
@@ -40,16 +40,12 @@ func Test_Workflow(t *testing.T) {
 
 	env.ExecuteWorkflow(PSOWorkflow, "sphere")
 
-	//env.AssertExpectations(t) // assert any OnActivity and Return
 	require.True(t, env.IsWorkflowCompleted())
-	//require.NoError(t, env.GetWorkflowError())
 	require.Equal(t, env.GetWorkflowError().Error(), "ContinueAsNew") // consider recreating a new test env on every iteration and calling execute workflow with the arguments from the previous iteration (contained in ContinueAsNewError)
-	//require.Equal(t, "evaluateFitnessActivity", activityCalled) //activityCalled is a vector with many activities called
-	queryAndVerify(t, env, "ContinueAsNew issued")
```
Comment (marcofiocco) [inline cmd/samples/pso/workflow_test.go]: Cannot see how to run a query for a child workflow here

File: cmd/samples/pso/workflow.go (hunk 9 above)
Comment (mfateev) [reply cmd/samples/pso/workflow.go]: ExecuteChildWorkflow returns ChildWorkflowFuture. Its GetChildWorkflowExecution method returns a Future that contains WorkflowExecution structure that contains runID. This future becomes ready as soon as the child is started.

```
		childWorkflowFuture := workflow.ExecuteChildWorkflow(ctx, child, ...)
		var childWE WorkflowExecution
		err := childWorkflowFuture.GetChildWorkflowExecution().Get(&childWE);
                 if err != nil {
                      return err
                 }
                 runID = childWE.RunID

```

Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: I get an error `undefined: WorkflowExecution`, where it is supposed to be defined?

File: cmd/samples/pso/workflow.go [hunk 17]
```diff
@@ -89,11 +89,11 @@ func PSOWorkflow(ctx workflow.Context, functionName string) (err error) {
 		}
 
 		var goalReached bool
-		childFuture := workflow.ExecuteChildWorkflow(ctx, PSOChildWorkflow, *swarm, 1)
-		//var childWE WorkflowExecution
-		//childFuture.GetChildWorkflowExecution().Get(&childWE)
-		err = childFuture.Get(ctx, &goalReached)
-		//ctx = workflow.WithValue(ctx, QueryResultName, childWE.RunID)
+		childWorkflowFuture := workflow.ExecuteChildWorkflow(ctx, PSOChildWorkflow, *swarm, 1)
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: It looks like the childWorklfowID is the same for each iteration. I would generate a different one for each for clarity and troubleshooting.

File: cmd/samples/pso/workflow.go [hunk 18]
```diff
@@ -58,16 +58,16 @@ func PSOWorkflow(ctx workflow.Context, functionName string) (err error) {
...
 		ExecutionStartToCloseTimeout: time.Minute,
 	}
 	ctx = workflow.WithChildOptions(ctx, cwo)
 
-	// Setup query handler for query type "stage"
-	ctx = workflow.WithValue(ctx, QueryResultName, childID)
+	// Setup query handler for query type "child"
+	ctx = workflow.WithValue(ctx, QueryResultName, childWorkflowID) // Don't need the runID when querying. Use only workflowID and it is going to query the latest one
 	err = workflow.SetQueryHandler(ctx, "child", func(input []byte) (string, error) {
 		return ctx.Value(QueryResultName).(string), nil
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: Any reason that context value is used instead of just returning childWorklfowID variable?

File: cmd/samples/pso/workflow.go [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Child workflow execution started.")
+
+	// Set activity options
+	ctx = workflow.WithActivityOptions(ctx, ActivityOptions)
+
+	// Setup query handler for query type "iteration"
+	ctx = workflow.WithValue(ctx, QueryResultName, "started")
+	err := workflow.SetQueryHandler(ctx, "iteration", func(input []byte) (string, error) {
+		return ctx.Value(QueryResultName).(string), nil
```
Comment (mfateev) [inline cmd/samples/pso/workflow.go]: I don't think there is a need to use context value. Just create a variable and return it directly. See https://github.com/samarabbas/cadence-samples/blob/master/cmd/samples/recipes/query/query_workflow.go

File: cmd/samples/pso/workflow.go (hunk 17 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: Right, I'll leave Cadence to generate it for me then

File: cmd/samples/pso/workflow.go (hunk 18 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: I was just trying to show how to create a global variable that is context specific. But yeah, not strictly necessary in this case

File: cmd/samples/pso/workflow.go (hunk 19 above)
Comment (marcofiocco) [reply cmd/samples/pso/workflow.go]: OK

File: cmd/samples/pso/workflow_test.go [hunk 20]
```diff
@@ -39,14 +39,16 @@ func Test_Workflow(t *testing.T) {
...
-		queryAndVerify(t, env, "child", "PSO_child_workflow:default-test-run-id") // Child WorkflowID
+		childWorkflowID = workflowInfo.WorkflowExecution.ID
 	})
 
 	env.ExecuteWorkflow(PSOWorkflow, "sphere")
 
 	require.True(t, env.IsWorkflowCompleted())
-	//queryAndVerify(t, env, "child", "PSO_child_workflow:default-test-run-id_RunID") // Child WorkflowID+RunID
+	queryAndVerify(t, env, "child", childWorkflowID)
 	require.Equal(t, env.GetWorkflowError().Error(), "ContinueAsNew") // consider recreating a new test env on every iteration and calling execute workflow with the arguments from the previous iteration (contained in ContinueAsNewError)
```
Comment (marcofiocco) [inline cmd/samples/pso/workflow_test.go]: Is it possible to test a child workflow?
And is it possible to query a child workflow?
It seems to me that there is no support for it, am i missing something?

Comment (mfateev) [reply cmd/samples/pso/workflow_test.go]: I believe child workflow testing is supported. Have you tried querying it?

Comment (marcofiocco) [reply cmd/samples/pso/workflow_test.go]: Let me rephrase it, the child workflow returns a ContinueAsNewError that I can get with env.GetWorkflowError() but I cannot cast it to the structure to extract the data necessary to run another child iteration.

Comment (marcofiocco) [reply cmd/samples/pso/workflow_test.go]: also env.QueryWorkflow() is tied to the parent workflow as it does not recognise the query in the child. Should I create a new env for the child and call env.ExecuteWorkflow(PSOChildWorkflow, ...)?

Comment (mfateev) [reply cmd/samples/pso/workflow_test.go]: I see. I think we miss some features in child workflow testing.

**Comment by marcofiocco** [comment]
Awesome, thanks!

**Comment by dmetzgar** [comment]
@marcofiocco this repo is now under an Uber official group. Could you please update and sign the CLA? It looks like there are commits from Ali Hashemi that are not associated with a GitHub account. We will need him to sign the CLA as well.

**Comment by marcofiocco** [comment]
@dmetzgar thanks, all done! :)

**Comment by dmetzgar** [comment]
Thanks for the great contribution!
